#Laboratorio / Regresión Lineal

#Aprendizaje Automático

Alex Pérez
00329229

In [ ]:
import matplotlib.pyplot as plt
import numpy as np #matlab de python
from sklearn import datasets, linear_model
from sklearn.metrics import mean_squared_error, r2_score #MSE función de costo

# Load the diabetes dataset
diabetes_X_all, diabetes_y = datasets.load_diabetes(return_X_y=True)

#Visualizing the dataset
#import numpy as np
from tabulate import tabulate
headers = ['age','sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6', 'output']
complete_table = np.column_stack([diabetes_X_all, diabetes_y]) 
table = tabulate(complete_table[0:9,:], headers, tablefmt="fancy_grid")
#print ("Visualizing first 10 rows")
#print(table)

# **ACTIVIDAD**

Utilizando la base de datos del punto anterior, implemente un regresor lineal de múltiples variables. Use todas las características de la base de datos. Calcule el MSE y el score R2 en el conjunton de test. Use el mismo conjunto de entrenamiento y test del punto anterior.

In [ ]:
#Since we want to use all the variables, we do not reshape diabetes_X_all as we did
#when we used only one feature, feature 2 

#Everything else is performed in a similar way as before

# Split the data into training/testing sets
diabetes_X_train = diabetes_X_all[:-20] 
diabetes_X_test = diabetes_X_all[-20:] #last 20 for test

# Split the targets into training/testing sets
diabetes_y_train = diabetes_y[:-20]
diabetes_y_test = diabetes_y[-20:]

# Create linear regression object
regr = linear_model.LinearRegression()

# normalize data to have zero mean and std one (feature scaling)
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(diabetes_X_train)# scaler fit is just for training dataset 
diabetes_X_train_normalized = scaler.transform(diabetes_X_train) #normalize training set

# Train the model using the training sets
regr.fit(diabetes_X_train_normalized, diabetes_y_train) # 1. REPRESENTAR hypothesis, 2. EVALUAR MSE Error Cuadr[atico Medio 3. OPTIMIZAR (Gradient Descent)

# Make predictions using the testing set
#scaler.fit(diabetes_X_test) #EQUIVOCADO, NOOOO!!!
diabetes_X_test_normalized = scaler.transform(diabetes_X_test) #normalize test set (do not fit scaler again!)
diabetes_y_pred = regr.predict(diabetes_X_test_normalized)

# The coefficients theta_0 (inter.), theta_1 (pendiente)
print("Coefficients: \n", "theta_values = ", regr.coef_, " theta_0  = ", regr.intercept_)
#print("Coefficients: \n", regr.coef_, regr.intercept_)
# The mean squared error
print("Mean squared error (MSE Cost): %.2f" % mean_squared_error(diabetes_y_test, diabetes_y_pred))
# The coefficient of determination: 1 is perfect prediction, varía entre 0 y 1
print("Coefficient of determination: %.2f" % r2_score(diabetes_y_test, diabetes_y_pred))

#Since we are dealing with multiple variables, plotting the result is not a viable option


Coefficients: 
 theta_values =  [ 1.45070184e-02 -1.13011397e+01  2.42953223e+01  1.55991981e+01
 -3.86350110e+01  2.33851844e+01  4.81833365e+00  8.79950709e+00
  3.56450223e+01  3.60654891e+00]  theta_0  =  153.36255924170615
Mean squared error (MSE Cost): 2004.57
Coefficient of determination: 0.59


# **Actividad**

* **Mejoró el valor de R2 respecto del punto anterior? Razone el por qué.**

Anteriormente se obtuvo un $R^2$ de $0.47$. Al hacer un regresor lineal multivariable, el valor de $R^2$ es de $0.59$, lo que implica una mejora. Se aclara que es así puesto que mientras más cerca se esté de 1, mejor será el $R^2$. 

La razón de que ahora sea mejor se debe a que al hacer este regresor lineal se están teniendo en cuenta las $n$ variables $x$, por lo que se tiene en cuenta la influencia que tiene cada una al momento de determinar la salida, obteniendo una relación más cercana entre ellas.

* **Cuál es la característica más importante para determinar la salida? Por qué? (Tip: observe los valores de theta)**

Puesto que, observando los valores theta, el theta de la feature 2 es de $24.3$ aproximadamente, la característica más importante será la de **bmi**. Como aclaración, esto se debe a que su valor de theta es mayor que los valores de theta de todas las demás características. 